File: android/src/main/kotlin/live/hms/hmssdk_flutter/methods/HMSSessionStoreAction.kt
package live.hms.hmssdk_flutter.methods

import com.google.gson.JsonElement
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel.Result
import live.hms.hmssdk_flutter.HMSCommonAction
import live.hms.hmssdk_flutter.HMSErrorLogger
import live.hms.hmssdk_flutter.HMSExceptionExtension
import live.hms.hmssdk_flutter.HMSResultExtension
import live.hms.video.error.HMSException
import live.hms.video.sdk.HMSSessionMetadataListener
import live.hms.video.sessionstore.HmsSessionStore

class HMSSessionStoreAction {

    companion object {

        fun sessionStoreActions(call: MethodCall, result: Result, hmsSessionStore: HmsSessionStore?) {
            when (call.method) {
                "get_session_metadata_for_key" -> {
                    getSessionMetadataForKey(call, result, hmsSessionStore)
                }
                "set_session_metadata_for_key" -> {
                    setSessionMetadataForKey(call, result, hmsSessionStore)
                }
            }
        }

        /***
         * This is used to get session metadata corresponding to the provided key
         *
         * If the key is null we log the error and return from method since we have the let
         * block in place already
         *
         * This method returns [sessionMetadata] is the session metadata is available for corresponding key
         */
        private fun getSessionMetadataForKey(call: MethodCall, result: Result, hmsSessionStore: HmsSessionStore?) {
            val key = call.argument<String?>("key") ?: run {
                HMSErrorLogger.returnArgumentsError("key is null")
            }

            key?.let {
                key as String
                hmsSessionStore?.get(
                    key,
                    object : HMSSessionMetadataListener {
                        override fun onError(error: HMSException) {
                            result.success(HMSResultExtension.toDictionary(false, HMSExceptionExtension.toDictionary(error)))
                        }

                        override fun onSuccess(sessionMetadata: JsonElement?) {
                            var value : Any? = null

                            /**
                             * Here depending on the value we parse the JsonElement
                             * if it's a JsonPrimitive we parse it as String and then send to flutter
                             * if it's a JsonObject,JsonArray we convert it to String and then send to flutter
                             * if it's a JsonNull we send it as null
                             */
                            sessionMetadata?.let {
                                value = if(it.isJsonPrimitive){
                                    it.asString
                                } else if (it.isJsonNull){
                                    null
                                } else{
                                    it.toString()
                                }
                            }

                            result.success(HMSResultExtension.toDictionary(true, value))
                        }
                    },
                )
            }
        }

        /***
         * This is used to set session metadata corresponding to the provided key
         *
         * If the key is null we log the error and return from method since we have the let
         * block in place already
         *
         * This method sets the [data] provided during the method call
         * The completion of this method is marked by actionResultListener's [onSuccess] or [onError] callback
         */
        private fun setSessionMetadataForKey(call: MethodCall, result: Result, hmsSessionStore: HmsSessionStore?) {
            val key = call.argument<String?>("key") ?: run {
                HMSErrorLogger.returnArgumentsError("key is null")
            }

            val data = call.argument<String?>("data")

            key?.let {
                key as String
                hmsSessionStore?.set(data, key, HMSCommonAction.getActionListener(result))
            }
        }
    }
}

File: ios/Classes/Actions/HMSSessionStoreAction.swift
//
//  HMSSessionStoreAction.swift
//  hmssdk_flutter
//
//  Created by Yogesh Singh on 25/04/23.
//

import Foundation
import HMSSDK

class HMSSessionStoreAction {

    static func sessionStoreActions(_ call: FlutterMethodCall, _ result: @escaping FlutterResult, _ plugin: SwiftHmssdkFlutterPlugin) {

        switch call.method {
        case "get_session_metadata_for_key":
            getSessionMetadata(call, result, plugin)

        case "set_session_metadata_for_key":
            setSessionMetadata(call, result, plugin)

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    static private func getSessionMetadata(_ call: FlutterMethodCall, _ result: @escaping FlutterResult, _ plugin: SwiftHmssdkFlutterPlugin) {

        guard let store = plugin.sessionStore
        else {
            HMSErrorLogger.returnHMSException(#function, "Session Store is null", "NULL ERROR", result)
            return
        }

        guard let arguments = call.arguments as? [AnyHashable: Any],
            let key = arguments["key"] as? String
        else {
            HMSErrorLogger.returnArgumentsError("key is null")
            HMSErrorLogger.returnHMSException(#function, "Key to be fetched from Session Store is null.", "NULL ERROR", result)
            return
        }

        store.object(forKey: key) { value, error in

            if let error = error {
                HMSErrorLogger.returnHMSException(#function, "Error in fetching key: \(key) from Session Store. Error: \(error.localizedDescription)", "Key Fetching error", result)
                return
            }

            if value is String? {
                result(HMSResultExtension.toDictionary(true, value))
            } else {
                HMSErrorLogger.returnHMSException(#function, "Session metadata type is not compatible, Please use String? type while setting metadata", "Type Incompatibility Error", result)
            }

        }

    }

    static private func setSessionMetadata(_ call: FlutterMethodCall, _ result: @escaping FlutterResult, _ plugin: SwiftHmssdkFlutterPlugin) {

        guard let store = plugin.sessionStore
        else {
            HMSErrorLogger.returnHMSException(#function, "Session Store is null.", "NULL ERROR", result)
            return
        }

        guard let arguments = call.arguments as? [AnyHashable: Any],
            let key = arguments["key"] as? String
        else {
            HMSErrorLogger.returnHMSException(#function, "Key for the object to be set in Session Store is null.", "NULL ERROR", result)
            return
        }

        guard let data = arguments["data"]
        else {
            HMSErrorLogger.returnHMSException(#function, "Data for the key - \(key) to be set in Session Store is null.", "NULL ERROR", result)
            return
        }

        store.set(data, forKey: key) { _, error in

            if let error = error {
                HMSErrorLogger.returnHMSException(#function, "Error in setting data: \(data) for key: \(key) to the Session Store. Error: \(error.localizedDescription)", "Key Error", result)
                return
            }
            result(nil)
        }
    }
}
